Modules/mathmodule.c

```diff
@@ -1653,25 +1655,85 @@ math_isqrt(PyObject *module, PyObject *n)
         return PyLong_FromLong(0);
     }
 
+    /* c = (n.bit_length() - 1) // 2 */
     c = _PyLong_NumBits(n);
     if (c == (size_t)(-1)) {
         goto error;
     }
     c = (c - 1U) / 2U;
 
-    /* s = c.bit_length() */
-    s = 0;
-    while ((c >> s) > 0) {
-        ++s;
+    /* Fast path: if c <= 31 then n < 2**64 and we can compute directly with
+       a fast, almost branch-free algorithm. */
+    if (c <= 31U) {
+        /* Convert and shift n to get a uint64_t m with 2**62 <= m < 2**64. */
+        n_shift = 31U - c;
+        m = (uint64_t)PyLong_AsUnsignedLongLong(n);
+        Py_DECREF(n);
+        if (m == (uint64_t)(-1) && PyErr_Occurred()) {
+            return NULL;
+        }
+        m <<= 2U * n_shift;
+
+        /* With m bounded, the shifts in the algorithm can be made explicit. */
+        u = 1U + (uint32_t)(m >> 62);
+        u = (u << 1) + (uint32_t)(m >> 59) / u;
+        u = (u << 3) + (uint32_t)(m >> 53) / u;
+        u = (u << 7) + (uint32_t)(m >> 41) / u;
+        /* The last step needs a 64-bit division */
+        u = (u << 15) + (uint32_t)((m >> 17) / u);
+        /* Adjust u if necessary. We test `u*u - 1 >= m` instead of the simpler
+           `u*u > m` to deal correctly with the corner case in which the final
+           addition above overflowed to give u = 0. */
+        u -= ((uint64_t)u * u - 1U >= m);
+
+        /* shift back and return */
+        return PyLong_FromUnsignedLong((unsigned long)(u >> n_shift));
     }
 
-    a = PyLong_FromLong(1);
+    /* Slow path: n >= 2**64. We perform the first five iterations in C integer
+       arithmetic, then switch to using Python long integers. */
+
+    /* Compute c.bit_length(); from n >= 2**64 it follows that this is >= 6. */
+    c_bit_length = 6;
+    while ((c >> c_bit_length) > 0U) {
+        ++c_bit_length;
+    }
+
+    /* Shift and convert n to get a uint64_t m with 2**62 <= m < 2**64. */
+    n_shift = c - 31U;
+    shift = PyLong_FromSize_t(2U * n_shift);
+    if (shift == NULL) {
+        goto error;
+    }
+    b = PyNumber_Rshift(n, shift);
+    Py_DECREF(shift);
+    if (b == NULL) {
+        goto error;
+    }
+    m = (uint64_t)PyLong_AsUnsignedLongLong(b);
+    Py_DECREF(b);
+    if (m == (uint64_t)(-1) && PyErr_Occurred()) {
+        return NULL;
+    }
+
+    /* The first 4 steps can be performed entirely in 32-bit arithmetic;
+        the last needs 64-bit arithmetic. */
+    u = 1U + (uint32_t)(m >> 62);
```

Comment:
Could this code be shared?
```
if (fastpath) {
    // init fast path
}
else {
    // init general path
}
// common code
if (fastpath) {
    // return fast path
}
// general path
```
With new shifting functions this may be clearer.

Reply:
Right, I considered this, but in the end I decided it was cleaner to keep the two paths separate.

I'll try the refactoring and see how it looks.

Reply:
Sharing is a bit ugly, because the last line has to be different in the two cases:
```
u = (u << 15) + (uint32_t)((m >> 17) / u);
```
versus
```
v = (u << 15) + ((m >> 17) / u);
```

Reply:
Why not to use always the second one? In any case you cast `u` to 64-bit for the final check. The overhead is unlikely noticeable.

Reply:
Updated to share via a helper function.